### PR TITLE
M1-4 functional specs - completed. 

### PR DESCRIPTION
**Pull request by wang2208**

### CODE REVIEW

File: twtr/grails-app/controllers/twtr/MessageController.groovy [hunk 1]
```diff
@@ -19,13 +19,15 @@ class MessageController extends RestfulController<Message> {
         def messageText = request.JSON.messageText
 
 
-        if(params.accountId != null)
-            tweeter = Account.findById(params.accountId)
+        if(params.accountId != null) {
+            if(params.accountId.isLong())
```
Comment (donaldsawyer) [inline twtr/grails-app/controllers/twtr/MessageController.groovy:23]: As we're cleaning up things, could this be: 
`if( params.accountId )`

File: twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesErrorsSpec.groovy [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+        error.message.contains('Not Acceptable')
+
+        where:
+        description                       | badMessage
+        'invalid message text - empty'    | ''
+        'invalid message text - too long' | 'a' * 50
+        'invalid message text - null'     | null
+    }
+
+    def 'validate that no messages were created for account with #badId'() {
```
Comment (donaldsawyer) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesErrorsSpec.groovy:154]: should this be goodId and not badId?

File: twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesErrorsSpec.groovy [hunk 3]
```diff
@@ -0,0 +1,175 @@
+package twtr.MessageFunctionalTests
+
+import grails.test.mixin.integration.Integration
+import groovyx.net.http.HttpResponseException
+import groovyx.net.http.RESTClient
+import spock.lang.*
+import geb.spock.*
+
+@Integration
+@Stepwise
```
Comment (donaldsawyer) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesErrorsSpec.groovy:10]: If you skip the first test and just add the account in the global setup() and do a delete in the global cleanup(), you don't need the `@Stepwise` tag.

File: twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+        response.data
+
+        when:
+        response = restClient.get(path: "/accounts/${goodId}")
+
+        then:
+        response.status == 200
+        response.data
+        response.data.id == goodId
+        response.data.handle == "$goodHandle$accountNumber"
```
Comment (donaldsawyer) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy:62]: nice!  I like this!

File: twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+        response.data
+        response.data.handle == "$goodHandle$accountNumber"
+        response.data.messageText == messageText
+
+        where:
+        description                   | message          | messageText
+        'valid message text - short'  | shortMessageText | shortMessageText
+        'valid message text - normal' | goodMessageText  | goodMessageText
+        'valid message text - long'   | longMessageText  | longMessageText
+    }
```
Comment (donaldsawyer) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy:82]: Why is the messageText column needed here?  It's the same as message.  You can just reuse message in the then: clause.

File: twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesErrorsSpec.groovy (hunk 2 above)
Comment (wang2208) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesErrorsSpec.groovy:154]: should be #goodId. I will update this.

File: twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy (hunk 5 above)
Comment (wang2208) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy:82]: message is the input to create messages and messageText is the output from response. Is that your question?  I want to make sure the response of post is correct.

Comment (donaldsawyer) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy:82]: not quite.  If the variables used for message and messageText are actually the same, you don't need separate columns for them.  So, the only columns you need are description & message and the last line can be `response.data.messageText == message`

Comment (wang2208) [inline twtr/src/integration-test/groovy/twtr/MessageFunctionalTests/CreateMessagesHappySpec.groovy:82]: Oh, right. Thanks.

**Pull request merged**